File: python/src/multiagent_boilerplate/harness/validate.py
"""Minimal structural validator for the flat JSON Schemas in specs/schemas.

Deliberately not a full JSON Schema engine (no $ref/oneOf resolution) - mirrors
harness/validate.ts in the TypeScript runtime so both runtimes enforce the same contracts
without pulling in a full validation library for three flat, mostly-one-level-deep schemas.
"""

from __future__ import annotations

from typing import Any

from .schemas import load_schema


class SchemaValidationError(Exception):
    def __init__(self, schema_file: str, path: str, message: str) -> None:
        super().__init__(f'schema validation failed for {schema_file} at "{path}": {message}')


def _matches_type(type_name: str, data: Any) -> bool:
    if type_name == "object":
        return isinstance(data, dict)
    if type_name == "array":
        return isinstance(data, list)
    if type_name == "string":
        return isinstance(data, str)
    if type_name == "integer":
        return isinstance(data, int) and not isinstance(data, bool)
    if type_name == "number":
        return isinstance(data, (int, float)) and not isinstance(data, bool)
    if type_name == "boolean":
        return isinstance(data, bool)
    if type_name == "null":
        return data is None
    return True


def _check(schema: dict, data: Any, path: str, schema_file: str) -> None:
    enum = schema.get("enum")
    if enum is not None and data not in enum:
        raise SchemaValidationError(schema_file, path, f"expected one of {enum!r}, got {data!r}")

    schema_type = schema.get("type")
    types = schema_type if isinstance(schema_type, list) else ([schema_type] if schema_type else [])
    if types and not any(_matches_type(t, data) for t in types):
        raise SchemaValidationError(schema_file, path, f"expected type {'|'.join(types)}, got {type(data).__name__}")

    if schema_type == "object" and isinstance(data, dict):
        for key in schema.get("required", []):
            if key not in data:
                raise SchemaValidationError(schema_file, path, f'missing required property "{key}"')
        for key, prop_schema in schema.get("properties", {}).items():
            if key in data:
                _check(prop_schema, data[key], f"{path}.{key}", schema_file)

    if schema_type == "array" and isinstance(data, list) and "items" in schema:
        for i, item in enumerate(data):
            _check(schema["items"], item, f"{path}[{i}]", schema_file)


def validate_against(schema_file: str, data: Any) -> None:
    _check(load_schema(schema_file), data, "$", schema_file)


def validate_agent_task(data: Any) -> None:
    validate_against("agent-task.schema.json", data)


def validate_artifact_ref(data: Any) -> None:
    validate_against("artifact-ref.schema.json", data)


def validate_trace_span(data: Any) -> None:
    validate_against("trace-span.schema.json", data)

File: typescript/src/harness/validate.ts
import { loadSchema } from "./schemas.js";

/**
 * Minimal structural validator for the flat JSON Schemas in specs/schemas - deliberately not
 * a full JSON Schema engine (no $ref/oneOf resolution). Three schemas, all flat objects with
 * at most one level of nesting; a hand-rolled ~60 line checker is simpler and has one fewer
 * dependency than pulling in a full validation library for this. If specs/schemas grows more
 * complex, revisit this call.
 */
export class SchemaValidationError extends Error {
  constructor(schemaFile: string, path: string, message: string) {
    super(`schema validation failed for ${schemaFile} at "${path}": ${message}`);
  }
}

type JsonSchema = {
  type?: string | string[];
  required?: string[];
  properties?: Record<string, JsonSchema>;
  items?: JsonSchema;
  enum?: unknown[];
};

function check(schema: JsonSchema, data: unknown, path: string, schemaFile: string): void {
  if (schema.enum && !schema.enum.includes(data)) {
    throw new SchemaValidationError(schemaFile, path, `expected one of ${JSON.stringify(schema.enum)}, got ${JSON.stringify(data)}`);
  }

  const types = Array.isArray(schema.type) ? schema.type : schema.type ? [schema.type] : [];
  if (types.length && !types.some((t) => matchesType(t, data))) {
    throw new SchemaValidationError(schemaFile, path, `expected type ${types.join("|")}, got ${typeof data}`);
  }

  if (schema.type === "object" && data && typeof data === "object") {
    const obj = data as Record<string, unknown>;
    for (const key of schema.required ?? []) {
      if (!(key in obj)) throw new SchemaValidationError(schemaFile, path, `missing required property "${key}"`);
    }
    for (const [key, propSchema] of Object.entries(schema.properties ?? {})) {
      if (key in obj) check(propSchema, obj[key], `${path}.${key}`, schemaFile);
    }
  }

  if (schema.type === "array" && Array.isArray(data) && schema.items) {
    data.forEach((item, i) => check(schema.items as JsonSchema, item, `${path}[${i}]`, schemaFile));
  }
}

function matchesType(type: string, data: unknown): boolean {
  switch (type) {
    case "object":
      return typeof data === "object" && data !== null && !Array.isArray(data);
    case "array":
      return Array.isArray(data);
    case "string":
      return typeof data === "string";
    case "integer":
      return typeof data === "number" && Number.isInteger(data);
    case "number":
      return typeof data === "number";
    case "boolean":
      return typeof data === "boolean";
    case "null":
      return data === null;
    default:
      return true;
  }
}

const cache = new Map<string, JsonSchema>();
function getSchema(schemaFile: string): JsonSchema {
  let schema = cache.get(schemaFile);
  if (!schema) {
    schema = loadSchema(schemaFile) as JsonSchema;
    cache.set(schemaFile, schema);
  }
  return schema;
}

export function validateAgainst(schemaFile: string, data: unknown): void {
  check(getSchema(schemaFile), data, "$", schemaFile);
}

export const validateAgentTask = (data: unknown) => validateAgainst("agent-task.schema.json", data);
export const validateArtifactRef = (data: unknown) => validateAgainst("artifact-ref.schema.json", data);
export const validateTraceSpan = (data: unknown) => validateAgainst("trace-span.schema.json", data);
